File: java/19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// first solution
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = head;
        ListNode prev = null;
        int length = 0;
        int count = 1;
        while (head != null) {
            length++;
            head = head.next;
        }
        
        if(length == 1 && n == 1) return null;

        head = dummy;
        while (head != null) {
            if (count == length - n + 1) {
                if(count == 1) return head.next;
                prev.next = head.next;
                break;
            }
            prev = head;
            head = head.next;
            count ++;
        }

        return dummy;
    }
}

// two pointer solution
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = head;
        ListNode fast = dummy;
        ListNode slow = dummy;
        
        for(int i = 1; i <= n; i++) {
            fast = fast.next;
        }

        if(fast == null) return slow.next;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        slow.next = slow.next.next;

        return dummy;
    }
}

// cool solution
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        ListNode slow = dummy;
        dummy.next = head;
        
        for(int i = 1; head != null; head = head.next, i++) {
            if(i > n) slow = slow.next;
        }

        slow.next = slow.next.next;

        return dummy.next;
    }
}

File: golang/linkedList/19. Remove Nth Node From End of List.go
package linkedList

import (
	"fmt"
	"log"
	"strings"
)

// fast/slow pointer, TC: O(n), SC: O(1)
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	removed, fast := head, head
	for n > 0 {
		fast = fast.Next
		n--
	}

	var prev *ListNode = nil
	for fast != nil {
		prev = removed
		removed = removed.Next
		fast = fast.Next
	}

	if removed == head {
		return head.Next
	}
	prev.Next = removed.Next
	return head
}

func Test_removeNthFromEnd() {
	case1H5 := &ListNode{5, nil}
	case1H4 := &ListNode{4, case1H5}
	case1H3 := &ListNode{3, case1H4}
	case1H2 := &ListNode{2, case1H3}
	case1H1 := &ListNode{1, case1H2}

	ans1 := removeNthFromEnd(case1H1, 2)
	ans1Str := ""
	for ans1 != nil {
		ans1Str += fmt.Sprintf("%d -> ", ans1.Val)
		ans1 = ans1.Next
	}
	ans1Str = strings.TrimRight(ans1Str, " -> ")
	log.Printf(ans1Str)

	case2H1 := &ListNode{1, nil}

	ans2 := removeNthFromEnd(case2H1, 1)
	ans2Str := ""
	for ans2 != nil {
		ans2Str += fmt.Sprintf("%d -> ", ans2.Val)
		ans2 = ans2.Next
	}
	ans2Str = strings.TrimRight(ans2Str, " -> ")
	log.Printf(ans2Str)

	case3H2 := &ListNode{2, nil}
	case3H1 := &ListNode{1, case3H2}

	ans3 := removeNthFromEnd(case3H1, 1)
	ans3Str := ""
	for ans3 != nil {
		ans3Str += fmt.Sprintf("%d -> ", ans3.Val)
		ans3 = ans3.Next
	}
	ans3Str = strings.TrimRight(ans3Str, " -> ")
	log.Printf(ans3Str)

	case4H2 := &ListNode{2, nil}
	case4H1 := &ListNode{1, case4H2}

	ans4 := removeNthFromEnd(case4H1, 2)
	ans4Str := ""
	for ans4 != nil {
		ans4Str += fmt.Sprintf("%d -> ", ans4.Val)
		ans4 = ans4.Next
	}
	ans4Str = strings.TrimRight(ans4Str, " -> ")
	log.Printf(ans4Str)
}
